driver-async/src/main/com/mongodb/async/client/gridfs/GridFS.java

```diff
@@ -0,0 +1,205 @@
+                    chunkCollection.createIndex(new Document("files_id", 1).append("n", 1),
+                            new CreateIndexOptions().unique(true), new SingleResultCallback<Void>() {
+
+                        @Override
+                        public void onResult(final Void result, final Throwable t) {
+                            //TODO log
+                            if (t != null) {
+                                //failed to create index
+                            } else {
+                                // index ok
+                            }
+                        }
+                    });
+                }
+            }
+        });
+    }
+
+    /**
+     * Instantiate a new GridFSFile.
+     *
+     * @return GridFSFile
+     */
+    public GridFSFile create() {
+        GridFSFile gridFSFile = new GridFSFile(chunkSize);
+        gridFSFile.setId(new ObjectId());
+        gridFSFile.setUploadDate(new Date());
+        return injectContext(gridFSFile);
+    }
+
+    /**
+     * Find a GridFSFile by its id.
+     *
+     * @param id the id as ObjectId
+     * @param callback callback that is completed when the file is available
+     */
+    public void find(final ObjectId id, final SingleResultCallback<GridFSFile> callback) {
+        fileCollection.find(new Document("_id", id)).first(new SingleResultCallback<GridFSFile>() {
+            @Override
+            public void onResult(final GridFSFile result, final Throwable t) {
+                if (t != null) {
+                    callback.onResult(null, t);
+                } else {
+                    callback.onResult(injectContext(result), null);
+                }
+            }
+        });
+    }
+
+    /**
+     * List all GridFSFiles, restricted by the provided query.
+     * @param query a document to limit the results that are returned
+     * @param callback callback that will be completed when the result is available
+     */
+    public void listFiles(final Document query, final SingleResultCallback<List<GridFSFile>> callback) {
+        fileCollection.find(query)
+                .sort(new Document("filename", 1))
+                .into(new ArrayList<GridFSFile>(), new SingleResultCallback<ArrayList<GridFSFile>>() {
+
+            @Override
+            public void onResult(final ArrayList<GridFSFile> result, final Throwable t) {
+                if (t != null) {
+                    callback.onResult(null, t);
+                } else {
+                    for (GridFSFile file : result) {
+                        injectContext(file);
+                    }
+                    callback.onResult(result, null);
+                }
+            }
+        });
+    }
+
+    /**
+     * Delete a GridFSFile by its id.
+     *
+     * @param id the id as ObjectId
+     * @param callback callback that is completed when the file is deleted
+     */
+    public void delete(final ObjectId id, final SingleResultCallback<DeleteResult> callback) {
```

Comment:
Since we know we're deleting a single file, DeleteResult is not useful.  Better to return Void